Refactor dpkg rules

r2d4 opened the pull request:
* Allows multiple sources on the dpkg() rule
Sources will be tried in the order they appear.  For different priority, split up your package bundles into different dpkg rules.

* Automatically renames packages that are not valid bazel rule names.
Prints out a warning with instructions on how to reference the resulting deb.  The best solution I can think of here is to eventually output a tarball of deb packages.  Then, the output wouldn't have to be the individual deb files (which for some packages won't be able to be referenced by bazel labels).

* Allows multiple packages to be part of the same dpkg() rule

* Bumps version of dpkg_parser.par to v0.2

Fixes https://github.com/GoogleCloudPlatform/distroless/issues/70
Partially addresses  #68

mattmoor commented on base/BUILD:
```diff
@@ -32,9 +32,9 @@ docker_build(
-        "@openssl//file:pkg.deb",
+        "@package_bundle//file:libc6.deb",
+        "@package_bundle//file:libssl1.0.0.deb",
+        "@package_bundle//file:openssl.deb",
```
This is cool.  I really like this model, it is basically what I'd like to see us do for `package.json` or `requirements.txt`, e.g. have a single `repository` and export multiple targets from within.

mattmoor commented on cc/BUILD:
```diff
@@ -7,7 +7,7 @@ docker_build(
-        "@libgcc1//file:pkg.deb",
-        "@libstdcpp6//file",
+        "@package_bundle//file:libgcc1.deb",
+        "@package_bundle//file:libstdc__6.deb",
```
I thought of another trick for how to handle invalid characters over the weekend, it is relevant here.

within `@package_bundle` generate a `.bzl` file, then you:
```python
load("@package_bundle//file:the.bzl", "packages")
```

Where `packages` is a dictionary of unescaped package names to `BUILD` target, then you replace this line with:
```python
   packages["libstdc++6"],
```

... and voila :)

This enables us to employ `base64` encoding or `sha256` to better avoid collisions.

@alexeagle has been battling something similar, but I don't believe this completely solves his problem.

r2d4 replied:
Nice idea!  I implemented it

mattmoor commented on package_manager/dpkg.bzl:
```diff
@@ -1,13 +1,24 @@
 """)
 
+  for pkg in repository_ctx.attr.packages:
+    if pkg.find("+") > 0:
```
@damienmg is there a better way to check if a string is a valid label?

r2d4 replied:
This check is unnecessary now that the user doesn't have to do any conversion on their end.

r2d4 commented:
In the latest commit I addressed some feedback as well as adding checksum validation for the Packages.gz file and individual .debs.  The builds should be reproducible and secure now.

mattmoor commented on package_manager/util_test.py:
```diff
@@ -0,0 +1,13 @@
+
+class TestUtil(unittest.TestCase):
+
+
```
nit: one blank line, we should add a pylint check to Travis... :-/

r2d4 replied:
Opened #73, will rebase and fix all the warnings on top of that once it gets in

r2d4 commented:
Fixes #68
Fixes #72

MarkusTeufelberger commented:
Please document how exactly one can generate the sha256 hash for the dpkg_src rule. It is a bit hidden in the code...

dlorenc commented on WORKSPACE:
```diff
@@ -23,73 +23,43 @@ dpkg_src(
 )
 
-# For the glibc base image.
 dpkg(
```
Maybe rename this rule to dpkg_list?

r2d4 replied:
renamed

dlorenc commented on package_manager/dpkg_parser.py:
```diff
@@ -17,22 +17,34 @@
 import io
```
It looks like we're not running pylint on this file, could you do that manually and open a bug to get it in travis?

r2d4 replied:
done - pylint was enabled, but after this PR, so i rebased

dlorenc commented on package_manager/dpkg_parser.py:
```diff
@@ -17,22 +17,34 @@
+
+import re
+
+from os import path
```
Just import os.

r2d4 replied:
done

dlorenc commented on package_manager/dpkg_parser.py:
```diff
@@ -42,31 +54,51 @@
-def download_package_list(mirror_url, distro, arch):
+    package_to_rule_map = {}
+    for pkg_name in packages.split(","):
+        found = False
```
You can do this with a for: else: loop. I find those clearer.

r2d4 replied:
done

r2d4 commented:
Done.  Added a README to the package manager folder and updated our main readme to point to that for extra info.
